fcrepo-http-api/src/main/java/org/fcrepo/http/api/services/SparqlTranslateVisitor.java

```diff
@@ -0,0 +1,185 @@
+        newUpdates.forEach(newRequest::add);
+        return newRequest;
+    }
+
+    /**
+     * Perform a translation of all the triples in an Update adding them to the internal list.
+     * @param update the update request to translate.
+     */
+    private void translateUpdate(final Update update) {
+        final List<Quad> sourceQuads;
+        if (update instanceof UpdateDeleteWhere) {
+            sourceQuads = ((UpdateDeleteWhere)update).getQuads();
+        } else {
+            sourceQuads = ((UpdateData) update).getQuads();
+        }
+        final List<Quad> newQuads = translateQuads(sourceQuads);
+        final Update newUpdate = makeUpdate(update.getClass(), newQuads);
+        newUpdates.add(newUpdate);
+    }
+
+    /**
+     * Perform a translation of all the triples in an UpdateModify request.
+     * @param update the update request to translate
+     */
+    private void translateUpdate(final UpdateModify update) {
+        final UpdateModify newUpdate = new UpdateModify();
+        final List<Quad> insertQuads = (update.hasInsertClause() ? translateQuads(update.getInsertQuads()) :
+                Collections.emptyList());
+        insertQuads.forEach(q -> newUpdate.getInsertAcc().addQuad(q));
+
+        final List<Quad> deleteQuads = (update.hasDeleteClause() ? translateQuads(update.getDeleteQuads()) :
+                Collections.emptyList());
+        deleteQuads.forEach(q -> newUpdate.getDeleteAcc().addQuad(q));
+
+        final Element where = update.getWherePattern();
+        newUpdate.setElement(where);
+        newUpdates.add(newUpdate);
+    }
+
+    /**
+     * Perform the translation to a list of quads.
+     * @param quadsList the quads
+     * @return the translated list of quads.
+     */
+    private List<Quad> translateQuads(final List<Quad> quadsList) {
+        final List<Quad> newQuads = new ArrayList<>();
+        for (final Quad q : quadsList) {
+            final Node subject = translateId(q.getSubject());
+            final Node object = translateId(q.getObject());
+            final Quad quad = new Quad(q.getGraph(), subject, q.getPredicate(), object);
+            LOGGER.trace("Translated quad is: {}", quad);
+            newQuads.add(quad);
+        }
+        return newQuads;
+    }
+
+    /**
+     * Quads insert/delete data statements don't contain variables and use QuadDataAcc to accumulate,
+     * insert {} delete {} where {} statements can't contain variables and use QuadAcc to accumulate. This function
+     * simplifies the creation of the eventual Update.
+     * @param updateClass the class of Update we are starting with.
+     * @param quadList the list of Quads to generate the above class with.
+     * @return a subclass of Update with the provided Quads.
+     */
+    private Update makeUpdate(final Class<? extends Update> updateClass, final List<Quad> quadList) {
+        try {
+            if (updateClass.equals(UpdateDeleteWhere.class)) {
```

Comment:
In the code that you were working on last week, there was a case for `if (updateClass.equals(UpdateModify.class))` here as well. Is that not needed?

Reply:
I don't think so because the above `private void translateUpdate(final UpdateModify update)` returns the correct `UpdateModify` object. This is used by the `private void translateUpdate(final Update update) {` one that returns one of 3 classes.